lib/resourcecontrol/resource_control.go

```diff
@@ -0,0 +1,179 @@
+package resourcecontrol
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+
+	"golang.org/x/sys/unix"
+)
+
+// NOTE: calling log.Fatal() inside a library is usually considered bad
+// practice, in this case it only happens in the child process so the library
+// user will not be affected.
+
+const (
+	execEnvVar      = "OVERSEER_EXEC"
+	cpuMaxEnvVar    = "OVERSEER_CPU_MAX"
+	memMaxEnvVar    = "OVERSEER_MEM_MAX"
+	ioMaxRbpsEnvVar = "OVERSEER_IO_MAX_RBPS"
+	ioMaxWbpsEnvVar = "OVERSEER_IO_MAX_WBPS"
+
+	controlSubtree = "overseer"
+)
+
+var (
+	errCgroupNotMounted = errors.New("cgroup2 is not mounted")
+)
+
+// TODO: add more limits
+type ResourceLimits struct {
+	CPUMax    string
+	MemMax    string
+	IOMaxRbps string
+	IOMaxWbps string
+}
+
+// init will look for a special environment variable that, if present, will
+// signal the need to set up the resource limits and call unix.Exec() to start
+// the limited command
+func init() {
+	if os.Getenv(execEnvVar) == execEnvVar {
+		// Unset custom environment variables
+		for _, v := range []string{execEnvVar, cpuMaxEnvVar, memMaxEnvVar, ioMaxRbpsEnvVar, ioMaxWbpsEnvVar} {
+			os.Unsetenv(v)
+		}
+
+		setResourceLimits()
+		// TODO: drop privileges
+
+		if err := unix.Exec(os.Args[1], os.Args[1:], os.Environ()); err != nil {
```

Comment:
Is there a reason to use `unix.Exec` over `cmd.Exec`?

Reply:
I think you meant `cmd.Command()`? The reason is that `unix.Exec()` will replace the calling process with the new process (keeping the same PID and without further execution of the caller), while `cmd.Command()` will just spawn a new child process.

If you meant `syscall.Exec()` then it's just because the `syscall` package is deprecated (although it will probably work just fine for this case).

Reply:
Sorry I got that backwards, meant `exec.Command` as used elsewhere.
> The reason is that unix.Exec() will replace the calling process with the new process (keeping the same PID and without further execution of the caller), while cmd.Command() will just spawn a new child process.

That is true but I was more asking is there an underlying reason you want to replace the calling process rather than spawn a child? Seems like either should work fine

Reply:
It's just to keep the process tree cleaner, and it's easier to debug as there is one less process to consider.